qiskit/pulse/builder.py

```diff
@@ -0,0 +1,1996 @@
+            default_transpiler_settings = {}
+        self._transpiler_settings = default_transpiler_settings
+
+        if default_circuit_scheduler_settings is None:
+            default_circuit_scheduler_settings = {}
+        self._circuit_scheduler_settings = default_circuit_scheduler_settings
+
+        if schedule is None:
+            schedule = Schedule()
+        # pulse.Schedule: Root program block
+        self._schedule = schedule
+
+        self.set_active_block(Schedule())
+
+    def __enter__(self) -> Schedule:
+        """Enter this builder context and yield either the supplied schedule
+        or the schedule created for the user.
+
+        Returns:
+            The schedule that the builder will build on.
+        """
+        self._backend_ctx_token = BUILDER_CONTEXTVAR.set(self)
+        self._default_alignment_context.__enter__()
+        return self._schedule
+
+    @_compile_lazy_circuit_before
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Exit the builder context and compile the built pulse program."""
+        self._default_alignment_context.__exit__(exc_type, exc_val, exc_tb)
+        self.compile()
+        BUILDER_CONTEXTVAR.reset(self._backend_ctx_token)
+
+    @property
+    def backend(self):
+        """Returns the builder backend if set.
+
+        Returns:
+            Optional[BaseBackend]: The builder's backend.
+        """
+        return self._backend
+
+    @property
+    def block(self) -> Schedule:
+        """Return the active block of this bulder."""
+        return self._block
+
+    @property
+    @_requires_backend
+    def num_qubits(self):
+        """Get the number of qubits in the backend."""
+        return self.backend.configuration().n_qubits
+
+    @property
+    def transpiler_settings(self) -> Mapping:
+        """The builder's transpiler settings."""
+        return self._transpiler_settings
+
+    @transpiler_settings.setter
+    @_compile_lazy_circuit_before
+    def transpiler_settings(self, settings: Mapping):
+        self._compile_lazy_circuit()
+        self._transpiler_settings = settings
+
+    @property
+    def circuit_scheduler_settings(self) -> Mapping:
+        """The builder's circuit to pulse scheduler settings."""
+        return self._circuit_scheduler_settings
+
+    @circuit_scheduler_settings.setter
+    @_compile_lazy_circuit_before
+    def circuit_scheduler_settings(self, settings: Mapping):
+        self._compile_lazy_circuit()
+        self._circuit_scheduler_settings = settings
+
+    @_compile_lazy_circuit_before
+    def compile(self) -> Schedule:
+        """Compile and output the built pulse program."""
+        # Not much happens because we currently compile as we build.
+        # This should be offloaded to a true compilation module
+        # once we define a more sophisticated IR.
+        program = self._schedule.append(self.block, mutate=True)
+        self.set_active_block(Schedule())
+        return program
```

Comment:
You could re-enter it if you had a handle, however, since this is private you would have to call `with pulse.build` again and this would create a new builder. `_schedule` is the schedule that is returned when the user calls `with pulse.build() as sched`. `block` is the active scheduling block for the current context. For example `with pulse.align_left()` creates and sets a new active block for the current context. When this context is exited the block is appended to the parent block automatically.